### PR TITLE
Feat ratelimiter

### PR DESCRIPTION
**Pull request by matoszz**
Adds a new generic rate limiter based on cloudflare's design

### CODE REVIEW

File: pkg/middleware/ratelimiter/examples/http_middleware.go/http_middleware.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package main
```
Comment (golanglemonade) [inline pkg/middleware/ratelimiter/examples/http_middleware.go/http_middleware.go:1]: Can you added `pkg/middleware/ratelimiter/examples/http_middleware.go/` to the sonar exclude list so it's not included in the checks?

File: pkg/middleware/ratelimiter/ratelimiter.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+// inspired by Cloudflare's approach: https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+// LimitStore is the interface that represents limiter internal data store
+type LimitStore interface {
+	// Inc increments current window limit counter for key
+	Inc(key string, window time.Time) error
+	// Get gets value of previous window counter and current window counter for key
+	Get(key string, previousWindow, currentWindow time.Time) (prevValue int64, currValue int64, err error)
```
Comment (golanglemonade) [inline pkg/middleware/ratelimiter/ratelimiter.go:14]: I see `Size` in the in memory store, should that  (`Size() int`) be included in the interface?

File: pkg/middleware/ratelimiter/map_limit_store.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	val        int64
+	lastUpdate time.Time
+}
+
+// NewMapLimitStore creates new in-memory data store for internal limiter data
+func NewMapLimitStore(expirationTime time.Duration, flushInterval time.Duration) (m *MapLimitStore) {
+	m = &MapLimitStore{
+		data:           make(map[string]limitValue),
+		expirationTime: expirationTime,
+	}
```
Comment (golanglemonade) [inline pkg/middleware/ratelimiter/map_limit_store.go:30]: This should return the interface type, that will ensure the type fully implements the interface:
```suggestion
// NewMapLimitStore creates new in-memory data store for internal limiter data
func NewMapLimitStore(expirationTime time.Duration, flushInterval time.Duration) LimitStore {
	m := &MapLimitStore{
		data:           make(map[string]limitValue),
		expirationTime: expirationTime,
	}
```

File: pkg/middleware/ratelimiter/map_limit_store.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+// Get gets value of previous window counter and current window counter
+func (m *MapLimitStore) Get(key string, previousWindow, currentWindow time.Time) (prevValue int64, currValue int64, err error) {
+	m.mutex.RLock()
+
+	defer m.mutex.RUnlock()
+
+	prevValue = m.data[mapKey(key, previousWindow)].val
+	currValue = m.data[mapKey(key, currentWindow)].val
+
+	return prevValue, currValue, nil
```
Comment (golanglemonade) [inline pkg/middleware/ratelimiter/map_limit_store.go]: nit: you defined these returns in the function signature, you can just do this instead:
```suggestion
	return 
```

File: pkg/middleware/ratelimiter/ratelimiter.go (hunk 2 above)
Comment (matoszz) [reply pkg/middleware/ratelimiter/ratelimiter.go:14]: `size` is a function that returns `return len(m.data)` so it shouldn't be a part of the interface

File: pkg/middleware/ratelimiter/map_limit_store.go (hunk 3 above)
Comment (matoszz) [reply pkg/middleware/ratelimiter/map_limit_store.go:30]: `MapLimitStore` is an option in-memory implementation where others could be used; returning LimitStore doesn't make sense in this instance

**Review by golanglemonade — APPROVED** [review]

**Pull request merged**